File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        vector<int> freq(256, -1); // ASCII chars ke liye last index
        int left = 0, ans = 0;

        for (int right = 0; right < n; right++) {
            char c = s[right];

            // Agar duplicate mila aur uska index current window ke andar hai
            if (freq[c] >= left) {
                left = freq[c] + 1;  // window shrink karo
            }

            freq[c] = right; // last index update
            ans = max(ans, right - left + 1); // max window size
        }
        return ans;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
            if(s.length()==0) return 0;
            int ans = 1;
            HashSet<Character> set = new HashSet<>();
            int left = 0, right = 0;
            while (right < s.length()) {
                if (set.contains(s.charAt(right))) {
                    set.remove(s.charAt(left));
                    left++;
                } else {
                    set.add(s.charAt(right));
                    right++;
                    int sz = set.size();
                    ans = Math.max(ans, sz);
                }
            }

            return ans;
        
    }
}